File: pachi/pachi-rust/src/ball_source.rs
use godot::{
    classes::{IMarker2D, Marker2D, node::PhysicsInterpolationMode},
    global::randf_range,
    prelude::*,
};

use crate::ball::Ball;

#[derive(GodotClass)]
#[class(init, base=Marker2D)]
pub struct BallSource {
    #[export]
    launch_velocity: Vector2,

    #[export]
    #[init(val = Vector2::new(10., 10.))]
    launch_jitter: Vector2,

    base: Base<Marker2D>,
}

#[godot_api]
impl IMarker2D for BallSource {
    // TODO: launch ball
}

#[godot_api]
impl BallSource {
    pub fn launch_existing_ball(&self, mut ball: Gd<Ball>, strength: f32) {
        let mut parent = self.base().get_parent().unwrap();
        let position = self.base().get_position();
        let prior_interpolation_mode = ball.get_physics_interpolation_mode();
        let jitter = Vector2::new(
            randf_range(-self.launch_jitter.x as f64, self.launch_jitter.x as f64) as f32,
            randf_range(-self.launch_jitter.y as f64, self.launch_jitter.y as f64) as f32,
        );
        ball.set_physics_interpolation_mode(PhysicsInterpolationMode::OFF);
        if ball.get_parent().is_some() {
            ball.reparent(&parent);
        } else {
            parent.add_child(&ball);
        }
        ball.set_position(position);
        ball.set_linear_velocity(self.launch_velocity * strength + jitter);
        ball.set_physics_interpolation_mode(prior_interpolation_mode);
    }
}

File: pachi.04/content/ball_source/BallSource.cs
using System.Diagnostics;
using Godot;

public partial class BallSource : Marker2D
{
    [Export]
    public Vector2 LaunchVelocity { get; set; }

    [Export]
    public Vector2 LaunchJitter { get; set; } = new Vector2(10, 10);

    public void LaunchExistingBall(Ball ball, float strength)
    {
        Debug.Assert(ball != null);
        Debug.Assert(ball.Tier != null);

        Node parent = GetParent();
        if (parent == null) return;

        Vector2 position = Position;
        PhysicsInterpolationModeEnum priorInterpolationMode = ball.PhysicsInterpolationMode;

        float jitterX = (float)GD.RandRange(-LaunchJitter.X, LaunchJitter.X);
        float jitterY = (float)GD.RandRange(-LaunchJitter.Y, LaunchJitter.Y);
        Vector2 jitter = new Vector2(jitterX, jitterY);

        ball.PhysicsInterpolationMode = PhysicsInterpolationModeEnum.Off;

        if (ball.GetParent() != null)
        {
            ball.Reparent(parent);
        }
        else
        {
            parent.AddChild(ball);
        }

        ball.Position = position;
        ball.LinearVelocity = LaunchVelocity * strength + jitter;
        ball.PhysicsInterpolationMode = priorInterpolationMode;
    }
}
